### DISCUSSION TITLE
Utility classes for color

### DISCUSSION THREAD

**Discussion by LeaVerou** [shoelace-style/webawesome#937] [Ideas & Suggestions]
As discussed in the Color meeting with @davegandy: We need a bunch of utility classes for certain color stylings, both semantic (e.g. brand, danger) as well as literal (e.g. blue, red). Here is some research on use cases, and some thoughts around a proposed syntax.

## Use cases

Looking at some use cases from current styles:

| Light mode | Dark mode |
|--------|--------|
| <img width="413" alt="image" src="https://github.com/user-attachments/assets/308f607f-502b-487e-9297-22dff598420f">  | <img width="404" alt="image" src="https://github.com/user-attachments/assets/ab1fb9a2-3ddd-4e2e-b006-412732ddfaf4"> |
| <img width="366" alt="image" src="https://github.com/user-attachments/assets/c18b0974-6e9a-4ff5-90c6-6afc1e5154ca"> | <img width="352" alt="image" src="https://github.com/user-attachments/assets/2f00dd86-4f1a-48fe-b384-4314b0e8bc7b"> | 

I also made [a demo to explore all 16 combinations](https://codepen.io/leaverou/pen/aberMvm/3262e69710ae5a7591683f6f1ed36957):

<img width="505" alt="image" src="https://github.com/user-attachments/assets/321c0b09-08b5-4a07-b15a-4f613063a1fe">

#### Colors per element

My initial thinking was to have classes that each set a specific thing (background, foreground, accent color, border color) to the corresponding color (e.g. `bg-brand`, `fg-blue` etc). 

However, the common case **by far** is to use the same color for the whole element, and what varies is which parts are colored (background, foreground, border etc) and their attention level (loud, normal, quiet). So a naming scheme that would require repeating the color name for each part (e.g. `fg-brand bg-brand border-brand`) would get repetitive and tedious.

While for _most_ colors it doesn’t make sense to combine multiple colors on the same element, combining a color plus the neutral is not unheard of. E.g. right here on GitHub:

<img width="83" alt="image" src="https://github.com/user-attachments/assets/1c01e118-18e7-4f50-90be-90d28894153a">
<img width="84" alt="image" src="https://github.com/user-attachments/assets/b12be022-0c81-41a1-a96f-b301289bc900">

or from Slack’s context menu:
<img width="325" alt="image" src="https://github.com/user-attachments/assets/9fe2c2cc-4c52-4c84-b649-694b0d083942">
<img width="330" alt="image" src="https://github.com/user-attachments/assets/b35a8a6e-9f21-4c76-9354-b26bdac47d88">

Eliminating repetition is not just about the element itself. For elements that act as containers, their "core" color usually extends to descendants. E.g. think of a colored callout, with the same color buttons etc, or [a button group](https://webawesome-git-next-font-awesome.vercel.app/docs/components/button-group#theme-buttons)). Doing this should not require repeating the color for every descendant.

#### Outlines

Of the 12 outlined combinations, only 8 are useful:
- Loud fills don't combine with any border
- No/Quiet fills combine with all borders
- Normal fills combine with normal & loud borders

#### Other 

- In some cases, the color is used for other things too, e.g. a pulsating effect here in `<wa-badge>`: 
<img width="243" alt="image" src="https://github.com/user-attachments/assets/3b2aa7fd-a29c-4d5f-9d27-ae3a4a76684b">
- Text color is largely a derived variable and depends on background color

## Thoughts + strawman

These are some thoughts, ideas, and questions. The space is very complex and I’m sure this first stab is missing out on use cases that should really be covered.

### Inheritable classes to specify the "core" color

The use cases above make me think that we should design utility classes that make it possible to only specify the "core" color once (e.g. `color-brand`) unless you want to do something unusual. Text color would be derived from the other parameters (which also ensures a11y), again unless you’re doing something that requires more power. 

Each utility class would do two things: set the corresponding CSS properties (e.g. `background`, `color`, `accent-color` etc), as well as custom properties (e.g. `--wa-color-background`), to enable more complex use cases (e.g. add decorations in the same color as the background or border or do things like the pulse animation). 

With that in mind, I’m thinking, what if the "core" color is neutrals by default, but can be overridden by a `color-xxx` class, e.g. `color-brand` or `color-blue`. The class would inherit, so that until overridden again by descendants, anything within uses that set of tints as its core color. Alternatively, we could have `variant-xxx` for semantic names (e.g. `variant-success`) and `color-xxx` for literal hues (e.g. `color-blue`).

One thing I’m really struggling with: **Should these classes _do_ something by themselves, or just affect how *other* classes behave?** In the spirit of _incremental user effort should provide incremental value_, it feels like `<button class="variant-success">` should not require additional stuff to make it green, but if these classes also add styling, that could hurt their composability. Another possibility is to make this element-specific, but that could hurt its predictability.

If we decide that the color classes actually do affect styling, a modifier could be added to get the other behavior, i.e. _only_ set custom properties with no visible effect otherwise (e.g. `silent`? `quiet`? `sneaky`? 😁 ).

### Backgrounds & text colors

Depending on the answer to the question above, we could design the rest. E.g. one scheme might be:

- `fg`: Just a text color. 
- `bg`: Loud bg 
- `bg-tinted`: Normal bg
- `bg-subtle`: Quiet bg

These would come with suitable text colors + hover styles, the whole lot.

We could also have explicit interaction based utilities like `bg-hover`. Then for example, the GitHub cancel button would be:

```css
<button class="color-danger fg bg-hover"></button>
```

Though ideally we don't want users to have to also think about focus styles etc, so perhaps we need something higher level than that.

### Outlines

Some of this may indicate that we need to change our border tokens. Currently, the number of available borders for each fill is vastly different, and some combinations don't work at all (e.g. quiet border + loud fill). 

I wonder if the meaning of a quiet/normal/loud border should be sensitive to the fill, e.g. so that a loud fill has a different loud border than a normal fill and thus every combination works and looks good. E.g. currently it's not possible to have a "loud" fill with an outline of any kind, yet this is not unheard of (e.g. looking at this very form, the submit button falls in that category: 
<img width="165" alt="image" src="https://github.com/user-attachments/assets/ff701df5-8fb3-43e3-96e8-090caebcdcb0">)

In that case, we'd be able to split the outline out entirely into a set of composable utility classes like `outline-subtle`, `outlined`, `outline-loud` that can be used together with any utilities to specify fill/text colors (yes, I don't like the inconsistency either, any ideas? It feels wrong to call the middle one just `outline`, but `outlined-subtle` also feels wrong).

### Examples

This is succinct enough that it allows us to get rid of the current `variant` and `appearance` attributes throughout and just use the same utility classes all over. Here are some before and after examples (assuming color classes that _do_ affect styling, at least for certain elements):

```html
<wa-button variant="brand">Filled</wa-button>
<wa-button class="color-brand">Filled</wa-button>

<wa-button appearance="tinted" variant="brand">Filled</wa-button>
<wa-button class="color-brand bg-tinted">Filled</wa-button>

<wa-button appearance="outlined" variant="brand">Outlined</wa-button>
<wa-button class="color-brand outlined">Outlined</wa-button>

<wa-tag variant="neutral">Neutral</wa-tag>
<wa-tag class="color-neutral">Neutral</wa-tag>
```

Note that only the first one is actually shorter — and only if the button doesn't already have another class. If it does, even in that case it’s shorter:

```html
<wa-button variant="brand" class="foobar">Filled</wa-button>
<wa-button class="foobar color-brand">Filled</wa-button>
```

Another advantage of utility classes over per-component attributes is inheritance:

```html
<!-- Now: -->
<wa-callout variant="success">
  <wa-icon slot="icon" name="circle-check" variant="regular"></wa-icon>
  <p>Well done!</p>
  <wa-button variant="success" appearance="outlined">Cancel</wa-button>
  <wa-button variant="success">Next</wa-button>
</wa-callout>

<!-- After: -->
<wa-callout class="color-success">
  <wa-icon slot="icon" name="circle-check" variant="regular"></wa-icon>
  <p>Well done!</p>
  <wa-button class="outlined">Cancel</wa-button>
  <wa-button>Next</wa-button>
</wa-callout>
```

----

I'm still very unsure about all of this, I feel there are so many use cases out there I'm missing a ton. But I think, this could be big if we get it right — imagine if people could get the same effect with one or two readable WA classes that in Tailwind requires a long string of unreadable classes.

Also, should these classes be `wa-` prefixed? Let’s discuss this in https://github.com/shoelace-style/webawesome/discussions/224

**Comment by lindsaym-fa** [comment]
There are definitely a ton of different avenues for this... But I'm fairly aligned with what you have here (for now, at least — I've been arguing with myself a lot 😅).

> Should these classes do something by themselves, or just affect how other classes behave?

Absolutely the former, imo. I would say that our goal should be that, in most cases, using `color-*` alone yields a result a user can run with.

### re: Backgrounds & text colors

`fg`, `bg`, `bg-tinted`, and `bg-subtle` are giving me Bootstrap flashbacks. Rather than framing this as styling a specific property, what if we mirror our volume levels?
- `silent` (or maybe `whisper`?): a text-only appearance
- `quiet`: `*-fill-quiet` and `*-on-quiet`
- `normal`: `*-fill-normal` and `*-on-normal`
- `loud`: `*-fill-loud` and `*-on-loud`

### re: Outlines

> Currently, the number of available borders for each fill is vastly different, and some combinations don't work at all (e.g. quiet border + loud fill).
>
> I wonder if the meaning of a quiet/normal/loud border should be sensitive to the fill, e.g. so that a loud fill has a different loud border than a normal fill and thus every combination works and looks good.

The tangle of custom properties we'd have to introduce here makes me nervous. 😅 But, more importantly, I feel strongly that borders are just as independent as fills. You _can_ pair them together, but they aren't inextricably linked.

I wonder if the solution is really for me to reconsider how each of our `--wa-color-*-border-loud` properties are defined so that they're more compatible with `--wa-color-*-fill-loud`. i.e., each loud border is one step darker than the loud fill. The neutral loud fill would be an exception given it's so dark — having discernible border colors for a fill that dark would be tricky in any system.

That said, I agree that `outlined` is likely deserving of its own utility class. I think we could assume whether to use `*-border-quiet`, `*-border-normal`, or `*-border-quiet` based on the default component styles or the applied volume class.

Bring those two together might look like this:
w/o `outlined`
![fill only](https://github.com/user-attachments/assets/e09b1279-418f-47bd-a2c9-e0ce99c86899)
w/ `outlined`
![outlined](https://github.com/user-attachments/assets/6c7117ce-5ccb-486f-b334-58abedf82684)
(We likely also want a way to specify no border — i.e., `<wa-card>` has a border by default, but you want a borderless card.)

### Reservations and ramblings

In my work on design systems, my goal has always been to _constructively_ limit the choices a user can make to prevent poor combinations or inconsistency. Sanctioned design choices should be codified in the system. I'm generally wary of opening up too many ad hoc choices for the user.

That wariness turns me away from things like `bg-hover` and the set of `outline-quiet`, `outlined`, and `outline-loud`. In the example of GitHub's outlined cancel button, they likely _always_ want their `quiet` buttons to use a neutral background and border, variably colored text, and a matching hover background. We should have some way of building that into the system that doesn't rely on users making that choice every time via classes. And if that isn't a sanctioned design choice for a product, it shouldn't be easy for developers in that product to do.

I'm also a _tad_ hesitant asking users to make the choice between `quiet` and `normal` fills per component. With `appearance="filled | tinted | outlined | text"`, we (as design system authors) could decide whether `*-fill-quiet` or `*-fill-normal` would be more appropriate for the `tinted` appearance of a component — surfaces/containers like `<wa-callout>` and `<wa-card>` would use `*-fill-quiet`, and actionable components like `<wa-button>` and `<wa-tag>` would use `*-fill-normal`. I worry that, in users' hands, the choice between a quiet or normal fill in many cases will be arbitrary.

That doesn't change (yet, at least 😆) the fact that I'm on board with this path.

### What about `filled`?

Currently, form controls like `<wa-input>` and `<wa-select>` are outlined by default but support a `filled` attribute. I haven't thought through that too hard yet but wanted to flag it so that we can think about it more.

**Comment by LeaVerou** [comment]
Thank you for the thoughtful review @lindsaym-fa! I mostly agree with everything you wrote.

> The tangle of custom properties we'd have to introduce here makes me nervous. 😅 

For now, let's first focus on what the ideal API would be, since that is also currently unknown. Once that is known, we can examine how close we can get with reasonable implementation complexity, but if we block the design process by considering implementation complexity too early, we won't even know what the north star API is, so we won't be able to move towards it over time (not to mention that in some cases things that seem hard to implement at first turn out to be fairly easy, with or without a few tweaks).

> Absolutely the former, imo. I would say that our goal should be that, in most cases, using color-* alone yields a result a user can run with.

So rather than composable classes, you're proposing a single class name that encodes multiple things, right?
I think that’s a viable path, especially since @claviska feels strongly that utility classes should be `wa-` prefixed (#224) so having multiple classes would end up involving a lot of visual noise. 

> Rather than framing this as styling a specific property, what if we mirror our volume levels?

I like the idea of making it more "semantic" and higher-level, but something like `brand-fill-quiet` to me does not imply that it's higher level than `color-brand bg-quiet`, it seems like just a different syntax.

> In my work on design systems, my goal has always been to constructively limit the choices a user can make to prevent poor combinations or inconsistency. Sanctioned design choices should be codified in the system. I'm generally wary of opening up too many ad hoc choices for the user.

💯 completely agree!

Which is why I’m not so sure about:

> I feel strongly that borders are just as independent as fills. You can pair them together, but they aren't inextricably linked.
> 
> I wonder if the solution is really for me to reconsider how each of our `--wa-color-*-border-loud` properties are defined so that they're more compatible with `--wa-color-*-fill-loud`. i.e., each loud border is one step darker than the loud fill. The neutral loud fill would be an exception given it's so dark — having discernible border colors for a fill that dark would be tricky in any system.

If border colors are completely independent, there is no way to codify sanctioned design choices. No matter what we do, I don't think there's a way to (for example) design a quiet border that works together with a loud fill without any of them knowing about the other. And it also defeats the purpose of having semantic attention levels, since the same border color can be "loud" if used on no/quiet fills and "quiet" when used on loud fills. When it comes to borders (or even text) it seems to me that perhaps loudness is a property of **contrast**.

On a related note, there’s something else that has been bothering me: it seems to me that there is another thing at play here: the styling a given theme may want to assign to *every* element of a given type is a subset of what is possible. E.g. in some themes, primary buttons would be loud with a (subtle) outline, in others flat with no outline, in others normal with an outline, and in others normal without an outline. There is a similar variation for any of these containers: some pages may want secondary buttons to be outlined, while in others text-only. Some pages may want loud badges, others more subtle ones. 

Which makes me wonder if this is something that needs to be defined by the themer, and then simply have utility classes that apply these predefined combos.

_(I wrote this and then noticed you wrote exactly the same thing below:)_

> In the example of GitHub's outlined cancel button, they likely always want their quiet buttons to use a neutral background and border, variably colored text, and a matching hover background. We should have some way of building that into the system that doesn't rely on users making that choice every time via classes. And if that isn't a sanctioned design choice for a product, it shouldn't be easy for developers in that product to do.

OMG YES! So how do we make it happen?

> What about filled?

Yes, this should definitely be supported. Though again, I think this falls into what we're discussing above. Is there a reason to make just one input filled? It seems to me that this is more of a _style_ you'd want to apply to all of them.

----

Thinking some more wrt sanctioned design choices, it seems there are two things that are completely independent: color and style (background loudness, outline loudness, outline style etc), so it should be *possible* to set them independently. Being able to set color independently would also preserve the ability to specify color and have it inherit down to descendants. What if we have 
1. Classes that set the "core" color and leave style unaffected  
2. Classes that set the style and leave color unaffected
3. Classes that are shorthands and set both.

**Comment by lindsaym-fa** [comment]
> For now, let's first focus on what the ideal API would be

That's actually why I'm nervous about the additional custom properties required for multiple border volumes per fill: that's a part of our API for manual theming — i.e., how WA Free users (and WA Pro users who want to off-road) would create themes. That said...

> If border colors are completely independent, there is no way to codify sanctioned design choices. No matter what we do, I don't think there's a way to (for example) design a quiet border that works together with a loud fill without any of them knowing about the other. And it also defeats the purpose of having semantic attention levels, since the same border color can be "loud" if used on no/quiet fills and "quiet" when used on loud fills. When it comes to borders (or even text) it seems to me that perhaps loudness is a property of contrast.

...I am willing to concede on this point. But first, just to give a bit more color (ha) on my viewpoint, I'll say two (long-winded) things:

1. re: choice — I'm not sure that most folks are making choices about outline intensity per fill. We'd be introducing an additional layer to each decision to outline a fill. The choice wouldn't be as simple as, "I want an outline on this fill," but "I want an outline _with this intensity_ on this fill." 
   1. Anecdotally, glancing across the web, most sites seem to use borders either in lieu of fills or paired with surface-like backgrounds (e.g., [HubSpot](https://www.hubspot.com), [Salesforce](https://www.salesforce.com), [Coda](https://coda.io/), [Apple](https://www.apple.com), [Kickstarter](https://www.kickstarter.com), [Vercel](https://vercel.com/), [Gusto](https://gusto.com), [Ikea](https://www.ikea.com/us/en/), [IBM](https://www.ibm.com/us-en)). Pairing outlines with loud fills doesn't really seem that common. As noted, it happens (hi, GitHub! and also Font Awesome!) but even then: are these examples using quiet, normal, and loud borders on their loud fills, or just a single intensity?
   2. Looking at a small sample of other design systems ([Nord](https://nordhealth.design/tokens/), [Atlassian](https://atlassian.design/components/tokens/all-tokens#color-border), [Elastic](https://eui.elastic.co/#/theming/borders#color), [Stacks](https://stackoverflow.design/product/foundation/color-fundamentals/#borders)), design system authors don't seem to be sanctioning different border intensities per fill. (Elastic is particularly fun here — there's only **_one_** sanctioned border color!)
2. re: contrast — I completely agree with this: "loudness is a property of contrast." My model is that the scale of loudness is relative to a product's surfaces, not relative to every possible color. If surfaces are your ambient noise, then the loudness scale for color dictates how much an element cuts through that noise. (We could argue that's too limited, but I'd also argue that might be the point.) imo, how much contrast your text and borders have with your surfaces should be handled by the theme builder rather than ad hoc choices. This would be the role of the contrast slider we've mentioned here and there.

Moving on to:
> So rather than composable classes, you're proposing a single class name that encodes multiple things, right?

Exactly. And...
> I like the idea of making it more "semantic" and higher-level, but something like brand-fill-quiet to me does not imply that it's higher level than color-brand bg-quiet, it seems like just a different syntax.

I'd actually simplify this further: either `wa-color-brand wa-quiet` or `wa-color-brand-quiet`. With these classes, you'd be choosing color and volume for the entire component rather than for a specific aspect of the component. With _I think_ agrees with...
> What if we have
Classes that set the "core" color and leave style unaffected
Classes that set the style and leave color unaffected
Classes that are shorthands and set both.

And I would add "Classes that set the volume," changing the intensity of the entire combination of colors. So, we could have:
1. Classes that set the "core" color, such as `wa-color-brand` or `wa-color-purple`
2. Classes that set the volume, such as `wa-loud`, `wa-normal`, and `wa-quiet`
3. Classes that set the style, such as `wa-outlined` or `wa-filled` (?)
4. Classes that are shorthands and set all, such as `wa-color-brand-loud-outlined` or `wa-color-brand-outlined` (the latter would perhaps use the default volume of the component)

Am I on the same page there, or did I stray too far?

**Comment by LeaVerou** [comment]
This is one of these problems where the more you think, the worse the problem gets, isn't it? 😅

Thinking some more:
- It feels that essentially these are actually design tokens, just higher level than colors, sizes, etc. That’s why we don't want users to create arbitrary combinations and stick to a few sanctioned ones. E.g. your design system may not include outlined buttons at all. Do we still want to support outlined in that case? Similarly, there are design systems where *every* button has some kind of outline, in which case you don't want people to be able to create non-outlined buttons.
- It seems that many (most?) design systems do not stick to one loud/normal/quiet style and use multiple alternate styles. The exact mechanics of how they pick among alternatives are not always obvious to me. E.g. what's the rationale for using a filled input instead of an outlined one, when your design system supports both? Or having both pill buttons and rounded ones? etc
- In the same spirit as _loudness is a property of contrast_, the *exact same* style can be loud or quiet depending on everything else. E.g. take these buttons as an example: <img width="345" alt="image" src="https://github.com/user-attachments/assets/7494de07-26b9-4f9e-9460-00ae53f5877d"> In a design system where primary buttons were using the outlined style and secondary buttons the text-only style, outlined is "loud". But in another where primary buttons are using the first (filled) style, outlined is "quiet".
- It seems like it could be useful to be able to use the styles of individual components as high-level utility classes. E.g. `<a class="wa-button"` would style a link like a button, whereas `<wa-badge class="wa-tag"></wa-badge>` would style a badge like a tag. 
- There are certain use cases where you **genuinely** want low level classes that *just* set the text color/background/fill/stroke, possibly with different ones on hover. The main one that springs to mind is icons, where the color is often completely arbitrary.

---

And some responses:

> That's actually why I'm nervous about the additional custom properties required for multiple border volumes per fill: that's a part of our API for manual theming — i.e., how WA Free users (and WA Pro users who want to off-road) would create themes.

Good point. Though I'd argue it's a lower-level API for power users, so in the spirit of _put the pain on those who can bear it_, we should prioritize designing a great high-level API first, then a good lower-level API, and then think of how to implement these [^1]. 

[^1]: As an aside, one of the holy grails I’ve been chasing for a while is a way to get color tokens in a relative way (i.e. something that would make this possible: `oklch(from var(--wa-color) var(--wa-darker-2));`). If we find a way to do that we could actually do this pretty easily.

>    1. Anecdotally, glancing across the web, most sites seem to use borders either in lieu of fills or paired with surface-like backgrounds (e.g., [HubSpot](https://www.hubspot.com), [Salesforce](https://www.salesforce.com), [Coda](https://coda.io/), [Apple](https://www.apple.com), [Kickstarter](https://www.kickstarter.com), [Vercel](https://vercel.com/), [Gusto](https://gusto.com), [Ikea](https://www.ikea.com/us/en/), [IBM](https://www.ibm.com/us-en)). Pairing outlines with loud fills doesn't really seem that common. As noted, it happens (hi, GitHub! and also Font Awesome!) but even then: are these examples using quiet, normal, and loud borders on their loud fills, or just a single intensity?
>    2. Looking at a small sample of other design systems ([Nord](https://nordhealth.design/tokens/), [Atlassian](https://atlassian.design/components/tokens/all-tokens#color-border), [Elastic](https://eui.elastic.co/#/theming/borders#color), [Stacks](https://stackoverflow.design/product/foundation/color-fundamentals/#borders)), design system authors don't seem to be sanctioning different border intensities per fill. (Elastic is particularly fun here — there's only **_one_** sanctioned border color!)

Do we have a Figma file or something where we collect screenshots of these kinds of examples in the wild? It could be incredibly useful, both for this discussion and many others.

From a quick look though, there *are* usually multiple borders per fill (or fills per border), but many of them are for hover/focus states.

> imo, how much contrast your text and borders have with your surfaces should be handled by the theme builder rather than ad hoc choices. This would be the role of the contrast slider we've mentioned here and there.

Yup, I'm leaning towards the same conclusion. Though I'm not sure we want a single contrast slider. E.g. how much contrast you want your typography to have (in terms of weights, font sizes) and how much contrast your want your colors to have is largely orthogonal. And even within color I'm not sure there is a single contrast. E.g. you may want very subtle borders but very large contrast between background & text color.
